### ISSUE TITLE
Splitting application across several files?

### ISSUE THREAD

**Issue by spitfiredd** [pywren/pywren-ibm-cloud#143]
How do I break up a single file application into several modules?

For example, say I have an application with the following structure

```
app/
    - __init__.py
    - file1.py
    - file2.py
```

In `file1.py` i have,

```
import pywren_ibm_cloud as pywren
from app.file2 import my_map_function, my_reduce_function

config = { # contetns omitted }

iterdata = [1, 2, 3, 4] 

pw = pywren.ibm_cf_executor(config=config)
pw.map_reduce(my_map_function, iterdata, my_reduce_function)
result = pw.get_result()
print(result)
```

In `file2.py` i have,

```
def my_map_function(x):
    return x + 7


def my_reduce_function(results):
    total = 0
    for map_result in results:
        total = total + map_result
    return total
```

If I run `python -m app.file1` I get the following error:

```
(.venv) D:\python\pywren-test>python -m app.file1
IBM Cloud Functions init for Namespace: spitfiredd@gmail.com_dev
IBM Cloud Functions init for Host: https://us-south.functions.cloud.ibm.com
IBM Cloud Functions init for Runtime: ibmfunctions/action-python-v3.7 - 256MB
IBM Cloud Functions executor created with ID 680d869e-8958
Executor ID 680d869e-8958 Uploading function and data - Total: 1.5KiB
Executor ID 680d869e-8958 Starting function invocation: my_map_function() - Total: 4 activations
Executor ID 680d869e-8958 Uploading function and data - Total: 6.7KiB
Executor ID 680d869e-8958 Starting function invocation: my_reduce_function() - Total: 1 activations
Executor ID 680d869e-8958 Getting results...

    0%|                                                                                                         | 0/1

Traceback (most recent call last):
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "D:\python\pywren-test\app\file1.py", line 20, in <module>
    result = pw.get_result()
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wren.py", line 391, in get_result
    WAIT_DUR_SEC=WAIT_DUR_SEC)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wren.py", line 327, in monitor
    pbar=pbar, THREADPOOL_SIZE=THREADPOOL_SIZE, WAIT_DUR_SEC=WAIT_DUR_SEC)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wait.py", line 83, in wait
    pbar=pbar)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wait.py", line 310, in _wait_storage
    pool.map(get_result, f_to_wait_on)
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 657, in get
    raise self._value
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wait.py", line 294, in get_result
    f.result(throw_except=throw_except, internal_storage=internal_storage)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\future.py", line 291, in result
    reraise(*self._traceback)
  File "D:\python\pywren-test\.venv\lib\site-packages\six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "/action/1/src/pywren_ibm_cloud/action/jobrunner.py", line 198, in run
  File "/action/1/src/pywren_ibm_cloud/action/jobrunner.py", line 134, in _unpickle_function
ModuleNotFoundError: No module named 'app'
```

How do I run an app that isn't just a single file?  Note that if this was all in the same file that the program runs.

**Comment by spitfiredd** [comment]
I seem to have gotten it to run, for this example, I will add additional module `file3.py`.

```
app/
    - __init__.py
    - file1.py
    - file2.py
    - file3.py
```

**file1.py**
```
import pywren_ibm_cloud as pywren
from app.file2 import Test


def run_app(model=Test):
    config = { # my config vars }
    iterdata = [1, 2, 3, 4]

    pw = pywren.ibm_cf_executor(config=config)
    # pw.map_reduce(my_map_function, iterdata, my_reduce_function)

    pw.map_reduce(Test.my_map_function, iterdata, Test.my_reduce_function)
    result = pw.get_result()
    return result
```

**file2.py**
```
class Test:
    @staticmethod
    def my_map_function(x):
        return x + 7

    @staticmethod
    def my_reduce_function(results):
        total = 0
        for map_result in results:
            total = total + map_result
        return total
```

**file3.py**
```
from app.file1 import run_app

results = run_app()
print(results)
```

This will work, however in `file2` if I change `file2` to just functions and not a class, e.g.
```
    def my_map_function(x):
        return x + 7

    def my_reduce_function(results):
        total = 0
        for map_result in results:
            total = total + map_result
        return total
```
and import those functions into `file1` it will not work.  

I am not sure why though? Are functions pickled differently that classes?

**Comment by JosepSampe** [comment]
Hi @spitfiredd, and thanks for your comments. 

Is there any reason in your code to do function imports with absolute paths? In the first case (file1.py + file2.py), can you try to import the methods with the relative path instead of the absolute one? That is, from: `from app.file2 import Test` to: `from file2 import Test`.

**Comment by spitfiredd** [comment]
I need to be able to run this from the root directory. So I apologize my project would look like

```
myproj/
    app/
        - __init__.py
        - file1.py
        - file2.py
```

I need to be able to run everything from `C:\myproj`

**Comment by spitfiredd** [comment]
If I am understanding this correctly, I will also need a copy of my `app` in the runtime as well?

**Comment by JosepSampe** [comment]
@spitfiredd I replicated your example with the intention of finding out what is happening, and in my case it works fine.

Structure of the project:
```
myproj/
    run.py
    app/
        - __init__.py
        - file1.py
        - file2.py
```

**run.py**
```
from app.file1 import run_app

results = run_app()
print(results)
```

**app/file1.py**
```
import pywren_ibm_cloud as pywren
from app.file2 import my_map_function, my_reduce_function


def run_app():
    iterdata = [1, 2, 3, 4]

    pw = pywren.ibm_cf_executor()
    pw.map_reduce(my_map_function, iterdata, my_reduce_function)
    result = pw.get_result()
    return result
```

**app/file2.py**
```
def my_map_function(x):
    return x + 7


def my_reduce_function(results):
    total = 0
    for map_result in results:
        total = total + map_result
    return total
```

Just out of curious, are you getting the same error as in the first post with this structure?
PyWren pickles all the imported packages that are not contained in the runtime, so it also pickles the app.

**Comment by spitfiredd** [comment]
Yes, I am getting the same error,

Here is the version info to make sure we're both using the same version, this was installed through pip so I am not on the latest development version.  Also, I am running this on Windows 10 Pro, Version 1803, OS Build 17134.765.

```
>>>import pywren_ibm_cloud
>>>pywren_ibm_cloud.__version__
'1.0.12'
```
Here is my most recent traceback with the same file struct as your above post.

```
(.venv) D:\python\pywren-test>python run.py
IBM Cloud Functions init for Namespace: spitfiredd@gmail.com_dev
IBM Cloud Functions init for Host: https://us-south.functions.cloud.ibm.com
IBM Cloud Functions init for Runtime: ibmfunctions/action-python-v3.7 - 256MB
IBM Cloud Functions executor created with ID eb5cb7ef-2bc7
Executor ID eb5cb7ef-2bc7 Uploading function and data - Total: 2.3KiB
Executor ID eb5cb7ef-2bc7 Starting function invocation: my_map_function() - Total: 4 activations
Executor ID eb5cb7ef-2bc7 Uploading function and data - Total: 7.5KiB
Executor ID eb5cb7ef-2bc7 Starting function invocation: my_reduce_function() - Total: 1 activations
Executor ID eb5cb7ef-2bc7 Getting results...

    0%|                                                                                                         | 0/1

Traceback (most recent call last):
  File "run.py", line 3, in <module>
    results = run_app()
  File "D:\python\pywren-test\app\file1.py", line 20, in run_app
    result = pw.get_result()
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wren.py", line 391, in get_result
    WAIT_DUR_SEC=WAIT_DUR_SEC)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wren.py", line 327, in monitor
    pbar=pbar, THREADPOOL_SIZE=THREADPOOL_SIZE, WAIT_DUR_SEC=WAIT_DUR_SEC)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wait.py", line 83, in wait
    pbar=pbar)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wait.py", line 310, in _wait_storage
    pool.map(get_result, f_to_wait_on)
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 657, in get
    raise self._value
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "c:\users\spitf\appdata\local\programs\python\python37\Lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\wait.py", line 294, in get_result
    f.result(throw_except=throw_except, internal_storage=internal_storage)
  File "D:\python\pywren-test\.venv\lib\site-packages\pywren_ibm_cloud\future.py", line 291, in result
    reraise(*self._traceback)
  File "D:\python\pywren-test\.venv\lib\site-packages\six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "/action/1/src/pywren_ibm_cloud/action/jobrunner.py", line 198, in run
  File "/action/1/src/pywren_ibm_cloud/action/jobrunner.py", line 134, in _unpickle_function
ModuleNotFoundError: No module named 'app'
```

**Comment by JosepSampe** [comment]
At the moment, last dev commit is version 1.0.12, so we are both using the same version. However, in my case I'm using Ubuntu 18.04.  I will try to run it in Windows, although it's very weird that it has a different behavior depending of the OS.

**Comment by JosepSampe** [comment]
@spitfiredd I'm preparing a Windows environment to test the app. Just to verify, can you also try to test it in an Ubuntu (linux) Environment? For example in a Virtual machine or in a WSL Ubuntu?

**Comment by spitfiredd** [comment]
@JosepSampe  I tested this on a docker image and I can confirm that it runs successfully.

Here is the build of my `Dockerfile`

```
FROM python:3.7-slim

WORKDIR /app

COPY ./requirements.txt requirements.txt  

RUN pip3 install -r requirements.txt

COPY . /app

CMD ["python", "run.py" ]  
```

Output of `docker run pywren:test`:

```
D:\python\pywren-test>docker run pywren:test
  100%|██████████| 1/1
IBM Cloud Functions init for Namespace: spitfiredd@gmail.com_dev
IBM Cloud Functions init for Host: https://us-south.functions.cloud.ibm.com
IBM Cloud Functions init for Runtime: ibmfunctions/action-python-v3.7 - 256MB
IBM Cloud Functions executor created with ID 21605784-e70b
Executor ID 21605784-e70b Uploading function and data - Total: 2.3KiB
Executor ID 21605784-e70b Starting function invocation: my_map_function() - Total: 4 activations
Executor ID 21605784-e70b Uploading function and data - Total: 7.5KiB
Executor ID 21605784-e70b Starting function invocation: my_reduce_function() - Total: 1 activations
Executor ID 21605784-e70b Getting results...


Executor ID 21605784-e70b Finished getting results
38
```

Let me know if you need any additional testing assistance.

**Comment by spitfiredd** [comment]
@JosepSampe 

I edited **`pywren_ibm_cloud/serialize/utils.py`**

```
def create_mod_data(mod_paths):

    module_data = {}
    # load mod paths
    for m in mod_paths:
        print(f"mod path: {m}")
        if os.path.isdir(m):
            files = glob2.glob(os.path.join(m, "**/*.py"))
            pkg_root = os.path.abspath(os.path.dirname(m))
        else:
            pkg_root = os.path.abspath(os.path.dirname(m))
            files = [m]
        for f in files:
            f = os.path.abspath(f)
            with open(f, 'rb') as file:
                mod_str = file.read()

            dest_filename = f[len(pkg_root)+1:]
            if os.name == 'nt':
                dest_filename = dest_filename.replace('\\', '/')
            module_data[dest_filename] = bytes_to_b64str(mod_str)

    return module_data
```

This works now. Do you want me to do a pull request?

**Comment by JosepSampe** [comment]
@spitfiredd cool!
@gilv fyi

**Comment by spitfiredd** [comment]
I somewhat cleaner solution is you use `pathlib.Path`, e.g.

```
if os.name == 'nt':
    dest_filename  = Path(dest_filename ).as_posix()
```

**Comment by JosepSampe** [comment]
@spitfiredd I included the fix in this pull request (#145), as finally it was only required to change one line of code. In any case, many thanks to find it :)
